aiida/engine/processes/workchains/workchain.py

```diff
@@ -124,17 +124,51 @@ def insert_awaitable(self, awaitable):
         :type awaitable: :class:`aiida.engine.processes.workchains.awaitable.Awaitable`
         """
         self._awaitables.append(awaitable)
+
+        # Already assign the awaitable itself to the location in the context container where it is supposed to end up
+        # once it is resolved. This is especially important for the `APPEND` action, since it needs to maintain the
+        # order, but the awaitables will not necessarily be resolved in the order in which they are added. By using the
+        # awaitable as a placeholder, in the `resolve_awaitable`, it can be found and replaced by the resolved value.
+        if awaitable.action == AwaitableAction.ASSIGN:
+            self.ctx[awaitable.key] = awaitable
+        elif awaitable.action == AwaitableAction.APPEND:
```

Comment:
I have a question here, it seems the functionality assign_ can be replaced by the append_, is there any special user scenario for `assign_`?

Reply:
What is the attribute 'key' of awaitable ? I can't find where it is initialized.

Reply:
The key is set in `construct_awaitables`:https://github.com/aiidateam/aiida-core/blob/c817021b73cc874a87348948b04eace2ba0900b6/aiida/engine/processes/workchains/workchain.py#L147
Basically it is the key that is used in `to_context` or `ToContext` and so is the key that should be used to store the resolved value in the `ctx`

Reply:
> I have a question here, it seems the functionality assign_ can be replaced by the append_, is there any special user scenario for `assign_`?

Why do you say so? If you use assign, it will be assigned to that key in the `ctx` overriding any value that may have been there. If you use `append` it will create a list or simply append the value if the key already contained a list.